src/realm/parser/query_builder.cpp

```diff
@@ -0,0 +1,895 @@
+    }
+    return value;
+}
+
+Timestamp get_timestamp_if_valid(int64_t seconds, int32_t nanoseconds) {
+    const bool both_non_negative = seconds >= 0 && nanoseconds >= 0;
+    const bool both_non_positive = seconds <= 0 && nanoseconds <= 0;
+    if (both_non_negative || both_non_positive) {
+        return Timestamp(seconds, nanoseconds);
+    }
+    throw std::runtime_error("Invalid timestamp format");
+}
+
+Timestamp from_timestamp_values(std::vector<std::string> const& time_inputs) {
+
+    if (time_inputs.size() == 2) {
+        // internal format seconds, nanoseconds
+        int64_t seconds = stot<int64_t>(time_inputs[0]);
+        int32_t nanoseconds = stot<int32_t>(time_inputs[1]);
+        return get_timestamp_if_valid(seconds, nanoseconds);
+    }
+    else if (time_inputs.size() == 6 || time_inputs.size() == 7) {
+        // readable format YYYY-MM-DD-HH:MM:SS:NANOS nanos optional
+        tm created;
+        created.tm_year = stot<int>(time_inputs[0]) - 1900; // epoch offset (see man mktime)
+        created.tm_mon = stot<int>(time_inputs[1]) - 1; // converts from 1-12 to 0-11
+        created.tm_mday = stot<int>(time_inputs[2]);
+        created.tm_hour = stot<int>(time_inputs[3]);
+        created.tm_min = stot<int>(time_inputs[4]);
+        created.tm_sec = stot<int>(time_inputs[5]);
+
+        int64_t seconds = platform_timegm(created); // UTC time
+        int32_t nanoseconds = 0;
+        if (time_inputs.size() == 7) {
+            nanoseconds = stot<int32_t>(time_inputs[6]);
+        }
+        return get_timestamp_if_valid(seconds, nanoseconds);
+    }
+    throw std::runtime_error("Unexpected timestamp format.");
+}
+
+// check a precondition and throw an exception if it is not met
+// this should be used iff the condition being false indicates a bug in the caller
+// of the function checking its preconditions
+#define precondition(condition, message) if (!REALM_LIKELY(condition)) { throw std::logic_error(message); }
+
+using KeyPath = std::vector<std::string>;
+KeyPath key_path_from_string(const std::string &s) {
+    std::stringstream ss(s);
+    std::string item;
+    KeyPath key_path;
+    while (std::getline(ss, item, '.')) {
+        key_path.push_back(item);
+    }
+    return key_path;
+}
+
+struct PropertyExpression
+{
+    std::vector<size_t> indexes;
+    std::function<Table *()> table_getter;
+    size_t col_ndx;
+    DataType col_type;
+
+    PropertyExpression(Query &query, const std::string &key_path_string)
+    {
+        KeyPath key_path = key_path_from_string(key_path_string);
+        TableRef cur_table = query.get_table();;
+        for (size_t index = 0; index < key_path.size(); index++) {
+            size_t cur_col_ndx = cur_table->get_column_index(key_path[index]);
+            StringData table_name = cur_table->get_name();
+
+            precondition(cur_col_ndx != realm::not_found,
+                         util::format("No property '%1' on object of type '%2'", key_path[index], table_name));
```

Comment:
We need to be using the class name rather than the table name in these exceptions, so the user doesn't see messages like "No property 'intCol' on object of type 'class_LinkTypesObject'".